### PR TITLE
Genmod rankscore analysis report

### PR DESCRIPTION
**Pull request by torbjorgen**
This PR contains documentation presenting the results for testing RankScore and RankScoreNormalized integrity and performance metrics of Genmod with respect to MutAcc variants.

Genmod source code used for testing is in this PR: https://github.com/Clinical-Genomics/genmod/pull/108

### CODE REVIEW

**Comment by torbjorgen** [comment]
Hi @jemten  and @henrikstranneheim 

This PR is blocked because I'm unable to reproduce the dataset with new, patched genmod version from PR https://github.com/Clinical-Genomics/genmod/pull/108.

I need a new dataset that's been produced with the above PR to rerun the tests for structural variants, to generate the plots in this PR.

Can you please help me out here? Much appreciated.

**Comment by jemten** [comment]
I can help out with that. I'll drop by your desk and you can show me what's needed

**Comment by torbjorgen** [comment]
Thanks @jemten for supplying the data!! :+1:

File: src/rdds/exploration_rankscore/assemble_unique_variant_id.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    :param df: Dataframe containing at least ID (current index), ALT, REF fields (lowercase)
+    :return: Reindexed DF
+    :raises ValueError: In case attempting to create duplicate UID
+    """
+    df = df.copy()  # Don't modify original DF
+    unique_ids: List[bytes] = []
+    ids = df.index.values
+    positions = df.pos.values
+    alts = df.alt.values
+    refs = df.ref.values
```
Comment (henrikstranneheim) [inline src/rdds/exploration_rankscore/assemble_unique_variant_id.py]: Add type hint

File: src/rdds/exploration_rankscore/assemble_unique_variant_id.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import pandas as pd
+from typing import List
+
+
+def assemble_unique_variant_id(pos: int,
+                               id: bytes,
+                               alt: bytes,
+                               ref: bytes) -> bytes:
+    """
```
Comment (henrikstranneheim) [inline src/rdds/exploration_rankscore/assemble_unique_variant_id.py]:
```suggestion
def assemble_unique_variant_id(chromosome_pos: int,
                               variant_id: bytes,
                               alt_allele: bytes,
                               ref_allele: bytes) -> bytes:
    """
```
I know this, but to someone new, but I think it  might be good for someone not familiar with the VCF format to be more explicit

File: src/rdds/exploration_rankscore/assemble_unique_variant_id.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    unique_ids: List[bytes] = []
+    ids = df.index.values
+    positions = df.pos.values
+    alts = df.alt.values
+    refs = df.ref.values
+    for pos, id, alt, ref in zip(positions, ids, alts, refs):
+        unique_id = assemble_unique_variant_id(pos=pos, id=id, alt=alt, ref=ref)
+        unique_ids.append(unique_id)
+    unique_ids_df = pd.DataFrame(data={'UID': unique_ids}, index=ids)
+    df = pd.concat((df, unique_ids_df), axis=1)
```
Comment (henrikstranneheim) [inline src/rdds/exploration_rankscore/assemble_unique_variant_id.py]: Is there more mnemonic name than the type of container?

File: src/tests/exploration_rankscore/test_assemble_unique_variant_id.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def test_assemble_unique_variant_id():
+    """
+    Test for creating a variant UID based on ID, ALT, REF VCF fields
+    """
+    # GIVEN a variant with POS, ID, ALT, and REF metadata
+    new_id = assemble_unique_variant_id(pos=1000, id=b'10_1032245_CA_C', alt=b'C', ref=b'CA')
+    # WHEN assembling a new UID, THEN expect it to be correct
+    assert new_id == b'1000-10_1032245_CA_C-alt-C-ref-CA'
```
Comment (henrikstranneheim) [inline src/tests/exploration_rankscore/test_assemble_unique_variant_id.py]:
```suggestion
    # WHEN assembling a new UID
    new_id = assemble_unique_variant_id(pos=1000, id=b'10_1032245_CA_C', alt=b'C', ref=b'CA')

   # THEN return a binary string
    assert new_id == b'1000-10_1032245_CA_C-alt-C-ref-CA'
```

File: src/tests/exploration_rankscore/test_assemble_unique_variant_id.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def test_assemble_unique_variant_ids_in_dataframe():
+    """
+    Test for updating a dataframe with unique IDs (UIDs)
+    """
+    # GIVEN a dataframe with some dummy variants
+    df = pd.DataFrame(data={'pos': [0, 1], 'alt': [b'C', b'A'], 'ref': [b'A', b'C']}, index=[b'id0', b'id0'])
+    df = dataframe_assemble_unique_variant_id(df=df)
+    # WHEN assembling, reindexing the DF based on new UID, THEN expect the UIDs to be correct
```
Comment (henrikstranneheim) [inline src/tests/exploration_rankscore/test_assemble_unique_variant_id.py]:
```suggestion
    # WHEN assembling, reindexing the DF based on new UID
    df = dataframe_assemble_unique_variant_id(df=df)

    # THEN  return data frame with 2 UIDs 
```

File: src/tests/exploration_rankscore/test_assemble_unique_variant_id.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+def test_duplicate_uid_error():
+    """
+    Test for making sure duplicate UIDs are caught.
+    """
+    # GIVEN a dataframe with two variants with identical ID, ALT, REF
+    df = pd.DataFrame(data={'pos': [0, 0], 'alt': [b'C', b'C'], 'ref': [b'A', b'A']}, index=[b'id0', b'id0'])
+    # WHEN reindexing, THEN expect this to be caught
+    with pt.raises(ValueError):
+        dataframe_assemble_unique_variant_id(df=df)
```
Comment (henrikstranneheim) [inline src/tests/exploration_rankscore/test_assemble_unique_variant_id.py]:
```suggestion
    # WHEN reindexing
    with pt.raises(ValueError):
        dataframe_assemble_unique_variant_id(df=df)
        
       # THEN an error should be raised
```

File: src/tests/exploration_rankscore/test_assemble_unique_variant_id.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import pandas as pd
+import pytest as pt
```
Comment (henrikstranneheim) [inline src/tests/exploration_rankscore/test_assemble_unique_variant_id.py:2]:
```suggestion
import pandas
import pytest
```

I think this improves readability and reduces cognitive load

**Review by henrikstranneheim — APPROVED** [review]
Awesome 👍 . Looks like they replicate beautifully 🎉

**Comment by torbjorgen** [comment]
* Updated according to @henrikstranneheim suggestions.

**Pull request merged**